File: impl/golang/internal/health/publisher.go
package health

import (
	"encoding/json"
	"fmt"
	"sync"

	"github.com/neuron-sdk/neuron-go-sdk/internal/keylib"
	"github.com/neuron-sdk/neuron-go-sdk/internal/topic"
)

// ValidateOutboundHeartbeat validates a HeartbeatPayload before publishing.
// Checks are performed in strict order (V-PUB-01..07) and the first failure
// is returned immediately.
//
// senderClock is the publisher's current timestamp (abstract uint64, typically
// Unix seconds). The shutdown sentinel (deadline==0) bypasses all delta checks.
func ValidateOutboundHeartbeat(payload HeartbeatPayload, senderClock uint64) error {
	// V-PUB-01: payload type must be "heartbeat".
	if payload.Type != PayloadTypeHeartbeat {
		return NewHealthError(ErrInvalidPayloadType,
			fmt.Sprintf("expected payload type %q, got %q", PayloadTypeHeartbeat, payload.Type))
	}

	// V-PUB-02: version must be the current supported version.
	if payload.Version != CurrentVersion {
		return NewHealthError(ErrUnsupportedVersion,
			fmt.Sprintf("expected version %q, got %q", CurrentVersion, payload.Version))
	}

	// V-PUB-03: shutdown sentinel (deadline==0) bypasses all delta checks.
	if payload.NextHeartbeatDeadline == ShutdownSentinel {
		return nil
	}

	// V-PUB-04: deadline must be strictly in the future.
	if payload.NextHeartbeatDeadline <= senderClock {
		return NewHealthError(ErrDeadlineNotFuture,
			fmt.Sprintf("nextHeartbeatDeadline %d is not after senderClock %d",
				payload.NextHeartbeatDeadline, senderClock))
	}

	// V-PUB-05: delta must be at least MinDeadlineDelta.
	delta := payload.NextHeartbeatDeadline - senderClock
	if delta < MinDeadlineDelta {
		return NewHealthError(ErrDeadlineTooSoon,
			fmt.Sprintf("delta %d is below minimum %d", delta, MinDeadlineDelta))
	}

	// V-PUB-06: delta must not exceed MaxDeadlineDelta.
	if delta > MaxDeadlineDelta {
		return NewHealthError(ErrDeadlineTooFar,
			fmt.Sprintf("delta %d exceeds maximum %d", delta, MaxDeadlineDelta))
	}

	// V-PUB-07: role must be a recognized NodeRole.
	if !ValidNodeRole(payload.Role) {
		return NewHealthError(ErrUnrecognizedRole,
			fmt.Sprintf("unrecognized role %q", payload.Role))
	}

	return nil
}

// TrimPayload progressively trims optional fields from a HeartbeatPayload
// to bring its serialized JSON size under maxSize.
//
// Trimming order (least-to-most important optional data):
//  1. Peers
//  2. Capabilities
//  3. Location
//
// If the mandatory-fields-only payload still exceeds maxSize, returns ErrPayloadTooLarge.
func TrimPayload(payload *HeartbeatPayload, maxSize int) error {
	// Check current size.
	data, err := json.Marshal(payload)
	if err != nil {
		return WrapHealthError(ErrPayloadTooLarge, "failed to serialize payload for size check", err)
	}
	if len(data) <= maxSize {
		return nil
	}

	// Trim peers first (least important optional field).
	payload.Peers = nil
	data, err = json.Marshal(payload)
	if err != nil {
		return WrapHealthError(ErrPayloadTooLarge, "failed to serialize payload after trimming peers", err)
	}
	if len(data) <= maxSize {
		return nil
	}

	// Trim capabilities next.
	payload.Capabilities = nil
	data, err = json.Marshal(payload)
	if err != nil {
		return WrapHealthError(ErrPayloadTooLarge, "failed to serialize payload after trimming capabilities", err)
	}
	if len(data) <= maxSize {
		return nil
	}

	// Trim location last.
	payload.Location = nil
	data, err = json.Marshal(payload)
	if err != nil {
		return WrapHealthError(ErrPayloadTooLarge, "failed to serialize payload after trimming location", err)
	}
	if len(data) <= maxSize {
		return nil
	}

	// Mandatory fields alone exceed the limit.
	return NewHealthError(ErrPayloadTooLarge,
		fmt.Sprintf("mandatory fields alone (%d bytes) exceed maximum size %d", len(data), maxSize))
}

// PublishHeartbeat validates, trims, signs, and publishes a heartbeat payload.
//
// Steps:
//  1. Validate with ValidateOutboundHeartbeat
//  2. Trim payload to fit adapter's MaxMessageSize
//  3. Serialize to JSON
//  4. Create signed TopicMessage via topic.NewTopicMessage
//  5. Publish via adapter.Publish with FireAndForget mode
//
// No retry is attempted on publish failure (FR-H25).
func PublishHeartbeat(
	payload HeartbeatPayload,
	key *keylib.NeuronPrivateKey,
	stdOutRef topic.TopicRef,
	adapter topic.TopicAdapter,
	senderClock uint64,
	sequenceNumber uint64,
) (topic.PublishResult, error) {
	// Step 1: Validate.
	if err := ValidateOutboundHeartbeat(payload, senderClock); err != nil {
		return topic.PublishResult{}, err
	}

	// Step 2: Trim to fit backend size limit.
	maxSize := int(adapter.MaxMessageSize())
	if err := TrimPayload(&payload, maxSize); err != nil {
		return topic.PublishResult{}, err
	}

	// Step 3: Serialize payload to JSON.
	payloadJSON, err := json.Marshal(payload)
	if err != nil {
		return topic.PublishResult{}, WrapHealthError(ErrPayloadTooLarge, "failed to serialize payload", err)
	}

	// Step 4: Create signed TopicMessage.
	msg, err := topic.NewTopicMessage(key, senderClock, sequenceNumber, payloadJSON)
	if err != nil {
		return topic.PublishResult{}, WrapHealthError(ErrInvalidPayloadType, "failed to create topic message", err)
	}

	// Step 5: Publish with FireAndForget.
	result, err := adapter.Publish(stdOutRef, msg, topic.PublishOpts{
		ConfirmationMode: topic.FireAndForget,
	})
	if err != nil {
		return topic.PublishResult{}, err
	}

	return result, nil
}

// ScheduleNextHeartbeat computes the wall-clock time at which the next heartbeat
// should be sent.
//
// If the publish result is confirmed and carries a consensus timestamp, that
// timestamp is used as the reference time. Otherwise, the submitWallClock
// (sender's local clock at submission time) is used.
//
// Returns referenceTime + chosenDelta.
func ScheduleNextHeartbeat(result topic.PublishResult, chosenDelta uint64, submitWallClock uint64) uint64 {
	var referenceTime uint64
	if result.Confirmed && result.ConsensusTimestamp != nil {
		referenceTime = *result.ConsensusTimestamp
	} else {
		referenceTime = submitWallClock
	}
	return referenceTime + chosenDelta
}

// HeartbeatPublisher manages rate-limited heartbeat publishing.
// It ensures that heartbeats are not published more frequently than MinDeadlineDelta.
//
// HeartbeatPublisher is safe for concurrent use.
type HeartbeatPublisher struct {
	key             *keylib.NeuronPrivateKey
	stdOutRef       topic.TopicRef
	adapter         topic.TopicAdapter
	lastPublishTime uint64
	sequenceNumber  uint64
	mu              sync.Mutex
}

// NewHeartbeatPublisher constructs a HeartbeatPublisher.
func NewHeartbeatPublisher(
	key *keylib.NeuronPrivateKey,
	stdOutRef topic.TopicRef,
	adapter topic.TopicAdapter,
) *HeartbeatPublisher {
	return &HeartbeatPublisher{
		key:       key,
		stdOutRef: stdOutRef,
		adapter:   adapter,
	}
}

// Publish validates and publishes a heartbeat, enforcing rate limiting.
// If senderClock - lastPublishTime < MinDeadlineDelta, the call is rejected
// with ErrRateLimited. On success, lastPublishTime is updated to senderClock.
//
// Concurrency note: The mutex is held for the entire publish operation, including
// the network call to the adapter. This serializes concurrent publish attempts.
// This is acceptable because rate-limiting already enforces MinDeadlineDelta (10s)
// gaps between publishes, making concurrent contention practically impossible.
func (p *HeartbeatPublisher) Publish(payload HeartbeatPayload, senderClock uint64) (topic.PublishResult, error) {
	p.mu.Lock()
	defer p.mu.Unlock()

	// Rate limit check: reject if publishing too frequently.
	if p.lastPublishTime > 0 && senderClock-p.lastPublishTime < MinDeadlineDelta {
		return topic.PublishResult{}, NewHealthError(ErrRateLimited,
			fmt.Sprintf("must wait at least %d seconds between publishes, elapsed %d",
				MinDeadlineDelta, senderClock-p.lastPublishTime))
	}

	p.sequenceNumber++
	result, err := PublishHeartbeat(payload, p.key, p.stdOutRef, p.adapter, senderClock, p.sequenceNumber)
	if err != nil {
		p.sequenceNumber--
		return topic.PublishResult{}, err
	}

	p.lastPublishTime = senderClock
	return result, nil
}

File: impl/typescript/src/health/publisher.ts
/**
 * Health publisher validation.
 *
 * Spec reference: 005 contracts/health-publisher.md
 *   - V-PUB-01: payload.type === "heartbeat"
 *   - V-PUB-02: payload.version is recognized (major === 1)
 *   - V-PUB-03: If deadline === 0n, return (shutdown bypass)
 *   - V-PUB-04: deadline > senderClock
 *   - V-PUB-05: delta >= MIN_DEADLINE_DELTA
 *   - V-PUB-06: delta <= MAX_DEADLINE_DELTA
 *   - V-PUB-07: role in VALID_ROLES
 *
 * senderClock is in seconds (same unit as deadline delta constants).
 * The HeartbeatPayload.nextHeartbeatDeadline is in nanoseconds,
 * so we convert to seconds before comparing with constants.
 */

import type { HeartbeatPayload } from './payload.js';
import { VALID_ROLES } from './types.js';
import { MIN_DEADLINE_DELTA, MAX_DEADLINE_DELTA } from './constants.js';
import {
  invalidPayloadType,
  invalidVersion,
  invalidDeadline,
  invalidRole,
} from './errors.js';

/** One second in nanoseconds. */
const NANOS_PER_SECOND = 1_000_000_000n;

/**
 * Publisher-side validation before signing a heartbeat.
 *
 * FR-H13: All V-PUB checks MUST pass before the payload is signed and published.
 * Throws a NeuronError on the first failing check.
 *
 * @param payload - The HeartbeatPayload to validate
 * @param senderClock - Publisher's current wall clock in nanoseconds
 * @throws NeuronError with HEALTH_INVALID_PAYLOAD_TYPE (V-PUB-01)
 * @throws NeuronError with HEALTH_INVALID_VERSION (V-PUB-02)
 * @throws NeuronError with HEALTH_INVALID_DEADLINE (V-PUB-04..06)
 * @throws NeuronError with HEALTH_INVALID_ROLE (V-PUB-07)
 */
export function validateOutboundHeartbeat(
  payload: HeartbeatPayload,
  senderClock: bigint,
): void {
  // V-PUB-01: type === "heartbeat"
  if (payload.type !== 'heartbeat') {
    throw invalidPayloadType(payload.type);
  }

  // V-PUB-02: version recognized (major === 1)
  const majorStr = payload.version.split('.')[0];
  if (majorStr === undefined || majorStr !== '1') {
    throw invalidVersion(payload.version);
  }

  // V-PUB-03: If deadline === 0n → bypass (shutdown sentinel)
  if (payload.nextHeartbeatDeadline === 0n) {
    return;
  }

  // V-PUB-04: deadline > senderClock
  if (payload.nextHeartbeatDeadline <= senderClock) {
    throw invalidDeadline('deadline must be in the future');
  }

  // Compute delta in seconds
  const deltaNanos = payload.nextHeartbeatDeadline - senderClock;
  const deltaSeconds = deltaNanos / NANOS_PER_SECOND;

  // V-PUB-05: delta >= MIN_DEADLINE_DELTA (10 seconds)
  if (deltaSeconds < MIN_DEADLINE_DELTA) {
    throw invalidDeadline(`deadline too soon: minimum delta is ${MIN_DEADLINE_DELTA.toString()} seconds`);
  }

  // V-PUB-06: delta <= MAX_DEADLINE_DELTA (86400 seconds)
  if (deltaSeconds > MAX_DEADLINE_DELTA) {
    throw invalidDeadline(`deadline too far: maximum delta is ${MAX_DEADLINE_DELTA.toString()} seconds`);
  }

  // V-PUB-07: role in VALID_ROLES
  if (!VALID_ROLES.includes(payload.role)) {
    throw invalidRole(payload.role);
  }
}
